examples/prodlda.py

```diff
@@ -0,0 +1,360 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Example: ProdLDA
+================
+In this example, we will follow [1] to implement the ProdLDA topic model from
+Autoencoding Variational Inference For Topic Models by Akash Srivastava and Charles
+Sutton [2]. This model returns consistently better topics than vanilla LDA and trains
+much more quickly. Furthermore, it does not require a custom inference algorithm that
+relies on complex mathematical derivations. This example also serves as an
+introduction to Flax and Haiku modules in NumPyro.
+
+For the interested reader, a nice extension of this model is the CombinedTM model [3]
+which utilizes a pre-trained sentence transformer (like https://www.sbert.net/) to
+generate a better representation of the encoded latent vector.
+
+**References:**
+    1. http://pyro.ai/examples/prodlda.html
+    2. Akash Srivastava, & Charles Sutton. (2017). Autoencoding Variational Inference
+       For Topic Models.
+    3. Federico Bianchi, Silvia Terragni, and Dirk Hovy (2021), "Pre-training is a Hot
+       Topic: Contextualized Document Embeddings Improve Topic Coherence"
+       (https://arxiv.org/abs/2004.03974)
+
+.. image:: ../_static/img/examples/prodlda.png
+    :align: center
+"""
+import argparse
+
+import matplotlib.pyplot as plt
+import pandas as pd
+from sklearn.datasets import fetch_20newsgroups
+from sklearn.feature_extraction.text import CountVectorizer
+from wordcloud import WordCloud
+
+import flax.linen as nn
+import haiku as hk
+import jax
+from jax import device_put, random
+import jax.numpy as jnp
+
+import numpyro
+from numpyro.contrib.module import flax_module, haiku_module
+import numpyro.distributions as dist
+from numpyro.infer import SVI, TraceMeanField_ELBO
+
+
+class HaikuEncoder:
+    def __init__(self, vocab_size, num_topics, hidden, dropout_rate):
+        self._vocab_size = vocab_size
+        self._num_topics = num_topics
+        self._hidden = hidden
+        self._dropout_rate = dropout_rate
+
+    def __call__(self, inputs, is_training):
+        dropout_rate = self._dropout_rate if is_training else 0.0
+
+        h = jax.nn.softplus(hk.Linear(self._hidden)(inputs))
+        h = jax.nn.softplus(hk.Linear(self._hidden)(h))
+        h = hk.dropout(hk.next_rng_key(), dropout_rate, h)
+
+        # NB: here we set `create_scale=False` and `create_offset=False` to reduce
+        # the number of learning parameters
+        h_mu = hk.Linear(self._num_topics)(h)
+        logtheta_loc = hk.BatchNorm(
+            create_scale=False, create_offset=False, decay_rate=0.99
```

Comment:
I think this is the same as the default in flax (though their parameter is called `momentum` rather than `decay_rate`).

Haiku requires that you set this, should I set it in the flax layers too to be more explicit? Also is `0.99` a good choice for this problem? I don't have that much intuition for batch norm...

Reply:
Good point! Luckily we didn't perform any prediction in this example, so the running mean/scale is not used anywhere. 

> intuition for batch norm

Currently, `momentum=0.1` is used by default in PyTorch [BatchNorm1d](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html#torch.nn.BatchNorm1d) (also in [ctm model](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/contextualized_topic_models/networks/decoding_network.py#L82)) to put less weights on the new input value when updating. The `momentum` notion in PyTorch is equivalent to `1 - momentum` notion in Flax/Haiku, so I think it is fine to use `0.9` here. I guess `0.99` also works but changing the default value in flax will make that hyperparameter more explicit to reader :D